### PR TITLE
fix Index parameters handling

### PR DESCRIPTION
**Pull request by alexanderguzhva**
* fixes https://github.com/milvus-io/milvus/issues/36980
* fixes https://github.com/milvus-io/milvus/issues/36920
* overrides parameter ranges to more practical
* more informative error messages 

/kind bug

### CODE REVIEW

**Comment by alexanderguzhva** [comment]
to all reviewers: It is expected that this commit triggers e2e problems, because it alters allowed parameter ranges to avoid internal Faiss exceptions (https://github.com/facebookresearch/faiss/blob/1a799d014b07aea1fca6788691078054f7865f4c/faiss/impl/ProductQuantizer.cpp#L64), which will eventually land into our codebase

File: src/index/ivf/ivf_config.h [hunk 1]
```diff
@@ -67,7 +67,8 @@ class IvfPqConfig : public IvfConfig {
     CFG_INT nbits;
     KNOHWERE_DECLARE_CONFIG(IvfPqConfig) {
         KNOWHERE_CONFIG_DECLARE_FIELD(m).description("m").for_train().set_range(1, 65536);
-        KNOWHERE_CONFIG_DECLARE_FIELD(nbits).description("nbits").set_default(8).for_train().set_range(1, 64);
+        // FAISS rejects nbits > 24, because it is not practical
+        KNOWHERE_CONFIG_DECLARE_FIELD(nbits).description("nbits").set_default(8).for_train().set_range(1, 24);
```
Comment (liliu-z) [inline src/index/ivf/ivf_config.h:71]: 1. Did you mean Faiss report error for `nbits > 24` and our E2E test has this case? Then why it didn't fail our E2E test?
2. Do we need to take care of the exist user? I prefer not.

Comment (alexanderguzhva) [reply src/index/ivf/ivf_config.h:71]: 1. this is what we'll need to migrate from the faiss baseline https://github.com/facebookresearch/faiss/blob/1a799d014b07aea1fca6788691078054f7865f4c/faiss/impl/ProductQuantizer.cpp#L64 eventually
2. NO USER uses even 16 bits

**Review by liliu-z — APPROVED** [review]
/lgtm

File: src/index/ivf/ivf_config.h (hunk 1 above)
Comment (liliu-z) [reply src/index/ivf/ivf_config.h:71]: > 1. this is what we'll need to migrate from the faiss baseline https://github.com/facebookresearch/faiss/blob/1a799d014b07aea1fca6788691078054f7865f4c/faiss/impl/ProductQuantizer.cpp#L64 eventually
> 2. NO USER uses even 16 bits

Then, looks like you need to modify `knowhere-test` or all other PR will suffer

**Comment by alexanderguzhva** [comment]
/hold
wait for https://github.com/milvus-io/milvus/pull/36844

**Comment by foxspy** [comment]
> /hold wait for [milvus-io/milvus#36844](https://github.com/milvus-io/milvus/pull/36844)

the milvus pr has been merged

**Comment by foxspy** [comment]
@alexanderguzhva 
e2e failed for ivfpq test



index = <knowhere.swigknowhere.IndexWrapFloat; proxy of <Swig Object of type 'IndexWrap< float > *' at 0x7f3e227cb510> >

xb = array([[ 0.8813573 , -0.12894475,  0.24229047, ...,  0.40410763,

        -1.4040895 ,  0.47813252],

       [ 0.2088593... ],

       [-0.18244962,  0.22511998, -0.6877572 , ..., -2.220745  ,

        -1.4776213 ,  0.61611724]], dtype=float32)

cfg = {'m': 1, 'metric_type': 'L2', 'nbits': 64, 'nlist': 128}

param_name = 'nbits', max_val = 64



    def check_build_param_max(index, xb, cfg, param_name, max_val):

        xb_dataset = knowhere.ArrayToDataSet(xb)

    

        cfg[param_name] = max_val + 1

        s = index.Build(xb_dataset, json.dumps(cfg))

        assert knowhere.Status(s) != knowhere.Status.success

    

        cfg[param_name] = max_val

        s = index.Build(xb_dataset, json.dumps(cfg))

>       assert knowhere.Status(s) == knowhere.Status.success

E       AssertionError: assert <Status.out_of_range_in_json: 3> == <Status.success: 0>

E        +  where <Status.out_of_range_in_json: 3> = <enum 'Status'>(3)

E        +    where <enum 'Status'> = knowhere.Status

E        +  and   <Status.success: 0> = <enum 'Status'>.success

E        +    where <enum 'Status'> = knowhere.Status



test_basic.py:171: AssertionError

----------------------------- Captured stderr call -----------------------------

I1028 14:27:46.748324  2738 index_factory.cc:55] [KNOWHERE][Create][python3] use key IVF_PQ_fp32 to create knowhere index IVF_PQ with version 5

E1028 14:27:46.748402  2738 config.h:322] [KNOWHERE][operator()][python3] Out of range in json: param 'm' (0) should be in range [1, 65536]

W1028 14:27:46.748466  2759 ivf.cc:369] [KNOWHERE][MatchNlist][knowhere_build3] nlist(128) is too large, adjust to a proper value

W1028 14:27:46.748476  2759 ivf.cc:371] [KNOWHERE][MatchNlist][knowhere_build3] Row num 20 match nlist 1

W1028 14:27:46.748478  2759 ivf.cc:380] [KNOWHERE][MatchNbits][knowhere_build3] nbits(8) is too large, adjust to a proper value

W1028 14:27:46.748481  2759 ivf.cc:390] [KNOWHERE][MatchNbits][knowhere_build3] Row num 20 match nbits 4

WARNING clustering 20 points to 1 centroids: please provide at least 39 training points

WARNING clustering 20 points to 16 centroids: please provide at least 624 training points

I1028 14:27:46.750362  2738 time_recorder.cc:49] [KNOWHERE][PrintTimeRecord][python3] Build index: done (1.910873 ms)

E1028 14:27:46.750453  2738 config.h:322] [KNOWHERE][operator()][python3] Out of range in json: param 'nbits' (0) should be in range [1, 24]

W1028 14:27:46.750530  2759 ivf.cc:369] [KNOWHERE][MatchNlist][knowhere_build3] nlist(128) is too large, adjust to a proper value

W1028 14:27:46.750540  2759 ivf.cc:371] [KNOWHERE][MatchNlist][knowhere_build3] Row num 20 match nlist 1

WARNING clustering 20 points to 1 centroids: please provide at least 39 training points

WARNING clustering 20 points to 2 centroids: please provide at least 78 training points

I1028 14:27:46.752362  2738 time_recorder.cc:49] [KNOWHERE][PrintTimeRecord][python3] Build index: done (1.847590 ms)

E1028 14:27:46.752442  2738 config.h:322] [KNOWHERE][operator()][python3] Out of range in json: param 'nbits' (65) should be in range [1, 24]

E1028 14:27:46.752494  2738 config.h:322] [KNOWHERE][operator()][python3] Out of range in json: param 'nbits' (64) should be in range [1, 24]

**Comment by alexanderguzhva** [comment]
@foxspy yes, we need https://github.com/milvus-io/knowhere-test/pull/206
Basically, we need to apply several patches at several locations at the same time

**Comment by alexanderguzhva** [comment]
superseded by #913

**Pull request closed without merging**